src/gui/sdlmain.cpp

```diff
@@ -1924,23 +1936,61 @@ static bool DetectResizableWindow()
 #endif // C_OPENGL
 }
 
+static bool detect_shader_fixed_size()
+{
+#if C_OPENGL
+	if (sdl.desktop.want_type != SCREEN_OPENGL)
+		return false;
+
+	const std::string sname = get_glshader_value();
+	return (sname == "crt-fakelottes-flat" || sname == "crt-easymode-flat");
+#else
+	return false;
+#endif // C_OPENGL
+}
+
+static std::tuple<int, int> detect_window_size()
+{
+	SDL_Rect bounds;
+	SDL_GetDisplayBounds(sdl.display_number, &bounds);
+	constexpr int resolutions[][2] = {
+	        {1600, 1200},
+	        {1280, 960},
+	        {1024, 768},
+	        {800, 600},
+	};
+	for (const auto &wh : resolutions) {
+		if (bounds.w > wh[0] && bounds.h > wh[1])
+			return std::make_tuple(wh[0], wh[1]);
+	}
+	return std::make_tuple(640, 480);
+}
+
 static void SetupWindowResolution(const char *val)
 {
 	assert(sdl.display_number >= 0);
 	std::string pref = NormalizeConfValue(val);
 
 	if (pref == "default" || pref.empty()) {
 #if defined(LINUX)
-		sdl.desktop.want_resizable_window = DetectResizableWindow();
+		sdl.desktop.want_resizable_window = detect_resizable_window();
 #else
 		sdl.desktop.want_resizable_window = false;
 #endif
 		sdl.desktop.window.use_original_size = true;
+
+		if (!sdl.desktop.want_resizable_window &&
+		    detect_shader_fixed_size()) {
+			const auto wh = detect_window_size();
+			sdl.desktop.window.use_original_size = false;
+			sdl.desktop.window.width = std::get<0>(wh);
+			sdl.desktop.window.height = std::get<1>(wh);
```

Comment:
Using `SDL_rect`  (or a `Dims` object) would tidy this up too.

Reply:
SDL_Rect has 4 members though, can't be constructed out of 2 integers :( There's SDL_Point though.

Reply:
> SDL_Rect has 4 members though 

Indeed.. a simple Dims object would be perfect.

Reply:
I used `SDL_Point`; it has fields `x` and `y`, but it's more readable than a tuple anyway.

Reply:
Looks like a nice compromise! x and y are perfectly intuitive and still represent the axis nicely. Maybe we can use `Point` for more refactoring (instead of carrying around separate sizes or dimensions).